sdk/java-sdk-protobuf/src/main/java/kalix/javasdk/workflowentity/WorkflowEntity.java

```diff
@@ -269,18 +288,119 @@ public Optional<Step> findByName(String name) {
     }
 
+    /**
+     * Add step to workflow definition. Step name must be unique.
+     * @param step A workflow step
+     */
     public Workflow<S> addStep(Step step) {
+      addStepWithValidation(step);
+      return this;
+    }
+
+    /**
+     * Add step to workflow definition with a dedicated {@link RecoverStrategy}. Step name must be unique.
+     * @param step A workflow step
+     * @param recoverStrategy A Step recovery strategy
+     */
+    public Workflow<S> addStep(Step step, RecoverStrategy<?> recoverStrategy) {
+      addStepWithValidation(step);
+      stepConfigs.add(new StepConfig(step.name(), step.timeout(), Optional.of(recoverStrategy)));
+      return this;
+    }
+
+    private void addStepWithValidation(Step step) {
       if (uniqueNames.contains(step.name()))
         throw new IllegalArgumentException("Name '" + step.name() + "' is already in use by another step in this workflow");
 
       this.steps.add(step);
       this.uniqueNames.add(step.name());
-      return this;
     }
 
     public void forEachStep(Consumer<Step> stepConsumer) {
       steps.forEach(stepConsumer);
     }
+
+    /**
+     * Define a timeout for the duration of the entire workflow. When the timeout expires, the workflow is finished and no transitions are allowed.
+     * @param timeout Timeout duration
+     */
+    public Workflow<S> timeout(Duration timeout) {
+      this.workflowTimeout = Optional.of(timeout);
+      return this;
+    }
+
+    /**
+     * Define a failover step name after workflow timeout. Note that recover strategy for this step can set only the number of max retries.
+     * @param stepName A failover step name
+     * @param maxRetriesRecoverStrategy A recovery strategy for failover step.
+     */
+    public Workflow<S> failoverTo(String stepName, MaxRetriesRecoverStrategy maxRetriesRecoverStrategy) {
+      this.failoverToStepName = Optional.of(stepName);
+      this.failoverToRecoveryStrategy = Optional.of(maxRetriesRecoverStrategy);
+      return this;
+    }
+
+    /**
+     * Define a failover step name after workflow timeout. Note that recover strategy for this step can set only the number of max retries.
+     * @param stepName A failover step name
+     * @param stepInput A failover step input
+     * @param maxRetriesRecoverStrategy A recovery strategy for failover step.
+     */
+    public <I> Workflow<S> failoverTo(String stepName, I stepInput, MaxRetriesRecoverStrategy maxRetriesRecoverStrategy) {
+      this.failoverToStepName = Optional.of(stepName);
+      this.failoverToStepInput = Optional.of(stepInput);
+      this.failoverToRecoveryStrategy = Optional.of(maxRetriesRecoverStrategy);
+      return this;
+    }
+
+    /**
+     * Define a default timeout for the duration of the step. If not set, a default value of 5 seconds is used.
+     * Can be overridden with step configuration.
+     */
+    public Workflow<S> stepTimeout(Duration timeout) {
```

Comment:
I don't remember if we have discussed/said it before, but what about calling it `defaultSetTimeout`? 

Would it help to make it more clear that this is a settings that applies to all steps?